### PR TITLE
Order sudoers file lines by role name

### PR DESCRIPTION
**Pull request by lxea**
Fix for #23095

### CODE REVIEW

File: lib/services/role.go [hunk 1]
```diff
@@ -2703,9 +2703,19 @@ func (set RoleSet) EnhancedRecordingSet() map[string]bool {
...
 	groups := make(map[string]struct{})
-	sudoers := make(map[string]struct{})
+	var sudoers []string
 	serverLabels := s.GetAllLabels()
-	for _, role := range set {
+
+	roleSet := make([]types.Role, len(set))
+	copy(roleSet, set)
+	slices.SortStableFunc(roleSet, func(a types.Role, b types.Role) bool {
+		if cmp := strings.Compare(a.GetName(), b.GetName()); cmp == 0 || cmp == 1 {
```
Comment (LKozlowski) [inline lib/services/role.go:2712]:
```suggestion
		if cmp := strings.Compare(a.GetName(), b.GetName()); cmp != -1 {
```

**Review by LKozlowski — APPROVED** [review]

**Review by r0mant — COMMENTED** [review]
@lxea Can you include an update to the docs in this PR to add a note saying how sudoer lines will be ordered in our [Host User Creation](https://goteleport.com/docs/server-access/guides/host-user-creation/) guide?

File: lib/services/role.go [hunk 2]
```diff
@@ -2703,9 +2703,19 @@ func (set RoleSet) EnhancedRecordingSet() map[string]bool {
...
 	serverLabels := s.GetAllLabels()
-	for _, role := range set {
+
+	roleSet := make([]types.Role, len(set))
+	copy(roleSet, set)
+	slices.SortStableFunc(roleSet, func(a types.Role, b types.Role) bool {
+		if cmp := strings.Compare(a.GetName(), b.GetName()); cmp != -1 {
+			return false
+		}
+		return true
```
Comment (smallinsky) [inline lib/services/role.go:2715]:
```suggestion
                return strings.Compare(a.GetName(), b.GetName()) == -1

```

**Review by smallinsky — CHANGES_REQUESTED** [review]

File: lib/services/role_test.go [hunk 3]
```diff
@@ -6060,6 +6066,108 @@ func TestHostUsers_HostSudoers(t *testing.T) {
...
+				Metadata: types.Metadata{
+					Name: "c",
+				},
+				Spec: types.RoleSpecV6{
+					Options: types.RoleOptions{
+						CreateHostUser: types.NewBoolOption(true),
+					},
+					Allow: types.RoleConditions{
+						NodeLabels:  types.Labels{types.Wildcard: []string{types.Wildcard}},
+						HostSudoers: []string{"sudoers entry 4"},
```
Comment (smallinsky) [inline lib/services/role_test.go:6095]: Can we extend it to: 
```suggestion
						HostSudoers: []string{"sudoers entry 4", "sudoers entry 1"},
```
to test deduplications ?

File: lib/services/role.go [hunk 4]
```diff
@@ -2738,18 +2743,26 @@ func (set RoleSet) HostUsers(s types.Server) (*HostUsersInfo, error) {
...
+	outer:
+		for _, sudoer := range sudoers {
+			for _, deniedSudoer := range role.GetHostSudoers(types.Deny) {
+				if deniedSudoer == "*" {
+					finalSudoers = nil
+					break outer
+				}
+				if sudoer != deniedSudoer {
+					finalSudoers = append(finalSudoers, sudoer)
+				}
```
Comment (smallinsky) [inline lib/services/role.go:2757]: I think that sill this logic is buggy. The final result of `finalSudoers`  will contain a duplication if a sudoers entry is duplicated across different roles 

See: https://github.com/gravitational/teleport/pull/23760/files#diff-1d9446cc64e6fb429df775e3d773435dce3c1e02b49cb0a03ed43ee1d4519f9aR6095

**Review by smallinsky — CHANGES_REQUESTED** [review]

**Review by smallinsky — APPROVED** [review]

**Review by r0mant — APPROVED** [review]

**Pull request merged**